domain/topn_slow_query.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package domain
+
+import (
+	"container/heap"
+	"time"
+
+	"github.com/pingcap/tidb/util/execdetails"
+)
+
+// topNSlowQueries maintains a heap to store recent slow queries.
+// N = 30, recent = 7 days by default.
+type topNSlowQueries struct {
+	data   []*slowQueryInfo
+	topN   int
+	recent time.Duration
+	ch     chan *slowQueryInfo
+}
+
+func newTopNSlowQueries(topN int, recent time.Duration) *topNSlowQueries {
+	return &topNSlowQueries{
+		data:   make([]*slowQueryInfo, 0, topN),
+		topN:   topN,
+		recent: recent,
+		ch:     make(chan *slowQueryInfo, 1000),
+	}
+}
+
+func (q *topNSlowQueries) Close() {
+	close(q.ch)
+}
+
+func (q *topNSlowQueries) Len() int           { return len(q.data) }
+func (q *topNSlowQueries) Less(i, j int) bool { return q.data[i].duration < q.data[j].duration }
+func (q *topNSlowQueries) Swap(i, j int)      { q.data[i], q.data[j] = q.data[j], q.data[i] }
+
+func (q *topNSlowQueries) Push(x interface{}) {
+	q.data = append(q.data, x.(*slowQueryInfo))
+}
+
+func (q *topNSlowQueries) Pop() interface{} {
```

Comment:
Pop can only return the minimum duration query, how can we implement topn and how can we just peek the heap?

Reply:
Maybe a b-tree is better?

Reply:
The read operation is not implemented in this PR.
Reading doesn't need to call Pop()

Reply:
Read will not be a frequent operation, so we just copy on read.
Get top 3 query is easy, copy the origin heap, Pop Pop Pop.